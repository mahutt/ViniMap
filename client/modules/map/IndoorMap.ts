import { IndoorMap, IndoorMapGeoJSON } from './Types';
import LocalLocations from '@/services/LocalLocations';

import hallGeojson from '@/assets/geojson/hall.json';
import jmsbGeojson from '@/assets/geojson/jmsb.json';
import VLandVEfloors from '@/assets/geojson/VLandVEfloors.json';
import websterGeojson from '@/assets/geojson/webster.json';
import evGeojson from '@/assets/geojson/ev.json';
import erGeojson from '@/assets/geojson/er.json';
import lsGeojson from '@/assets/geojson/ls.json';
import clGeojson from '@/assets/geojson/cl.json';
import fbGeojson from '@/assets/geojson/fb.json';
import nunsGeojson from '@/assets/geojson/nuns.json';
import vaGeojson from '@/assets/geojson/va.json';
import sdGeojson from '@/assets/geojson/sd.json';
import pcGeojson from '@/assets/geojson/pc.json';
import raGeojson from '@/assets/geojson/ra.json';
import psGeojson from '@/assets/geojson/ps.json';
import pyGeojson from '@/assets/geojson/py.json';
import rfGeojson from '@/assets/geojson/rf.json';
import ccGeojson from '@/assets/geojson/cc.json';
import adGeojson from '@/assets/geojson/ad.json';
import fcGeojson from '@/assets/geojson/fc.json';
import cjGeojson from '@/assets/geojson/cj.json';
import siGeojson from '@/assets/geojson/si.json';
import hcGeojson from '@/assets/geojson/hc.json';
import hbGeojson from '@/assets/geojson/hb.json';
import haGeojson from '@/assets/geojson/ha.json';
import huGeojson from '@/assets/geojson/hu.json';
import spGeojson from '@/assets/geojson/sp.json';
import geGeojson from '@/assets/geojson/ge.json';

import GeojsonService from '@/services/GeojsonService';
import type { Feature, Polygon } from 'geojson';
import { center } from '@turf/turf';

const getAllRooms = (geojson: IndoorMapGeoJSON): Feature<Polygon>[] => {
  const rooms: Feature<Polygon>[] = [];
  geojson.features.forEach((feature) => {
    if (feature.properties?.ref) {
      rooms.push(feature as Feature<Polygon>);
    }
  });
  return rooms;
};

const rawIndoorMaps = [
  {
    id: 'Hall Building (H)',
    mapboxIds: [22080570],
    geojson: hallGeojson,
    addr: '1455 Blvd. De Maisonneuve Ouest, Montreal, Quebec H3G 1M8',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Webster Library (LB)',
    mapboxIds: [
      1110900192, 1110900193, 1110900197, 1110900196, 1110900194, 1110900195, 1110900190,
      1110900191,
    ],
    geojson: websterGeojson,
    addr: 'Pavillion J.W. McConnell Bldg, 1400 Maisonneuve Blvd W, Montreal, Quebec H3G 1M8',
    opening_hours: '24/7',
  },
  {
    id: 'Engineering, CS, and VA Integrated Complex (EV)',
    mapboxIds: [1110145740, 1110145744, 1110145741, 356036803, 1110145743],
    geojson: evGeojson,
    addr: '1515 Saint-Catherine St W, Montreal, Quebec H3G 2H7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'ER Building (ER)',
    mapboxIds: [103896385],
    geojson: erGeojson,
    addr: '2155 Guy St, Montreal, Quebec H3H 2L9',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Learning Square (LS)',
    mapboxIds: [979438074],
    geojson: lsGeojson,
    addr: '1535 Blvd. De Maisonneuve Ouest, Montreal, Quebec H3G 1M9',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'John Molson Building (MB)',
    mapboxIds: [1109876433, 1109876436, 1109876434, 1109876435],
    geojson: jmsbGeojson,
    addr: '1600 Blvd. De Maisonneuve Ouest, Montreal, Quebec H3H 0A1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Faubourg Tower (FB)',
    mapboxIds: [103248058],
    geojson: fbGeojson,
    addr: '1600 Saint-Catherine St W, Montreal, Quebec H3H 2S7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Grey Nuns Building (GN)',
    mapboxIds: [17189298],
    geojson: nunsGeojson,
    addr: '1190 Guy St, Montreal, Quebec H3H 2L4',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'CL Building (CL)',
    mapboxIds: [103248055],
    geojson: clGeojson,
    addr: '1665 Saint-Catherine St W, Montreal, Quebec H3H 2S7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Visual Arts Building (VA)',
    mapboxIds: [103521746],
    geojson: vaGeojson,
    addr: '1395 René-Lévesque Blvd W, Montreal, Quebec H3G 2M5',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Vanier Library (VL / VE)',
    mapboxIds: [604324442, 604324441, 47332009, 604324443],
    geojson: VLandVEfloors,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: '24/7',
  },
  {
    id: 'Stinger Dome (SD)',
    mapboxIds: [129437736],
    geojson: sdGeojson,
    addr: '7200 Sherbrooke St W, Montreal, Quebec H4B 1R3',
    opening_hours: 'Mo-Su 09:00-22:00',
  },
  {
    id: 'Perform Center (PC)',
    mapboxIds: [604019520],
    geojson: pcGeojson,
    addr: '7200 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 06:30-22:00; Sa-Su 08:00-18:00',
  },
  {
    id: 'Recreation and Althletics Complex (RA)',
    mapboxIds: [47332005],
    geojson: raGeojson,
    addr: '7200 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 06:30-22:00; Sa-Su 08:00-18:00',
  },
  {
    id: 'Physical Services (PS)',
    mapboxIds: [47332006],
    geojson: psGeojson,
    addr: 'Montreal, Quebec H4B 2B9',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Psychology Building (PY)',
    mapboxIds: [17887805],
    geojson: pyGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Loyola Jesuit Hall and Conference Center (RF)',
    mapboxIds: [604324455],
    geojson: rfGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1E1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Central Building (CC)',
    mapboxIds: [604324456],
    geojson: ccGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1E1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Administration Building (AD)',
    mapboxIds: [604324457],
    geojson: adGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1E1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'F.C. Smith Building (FC)',
    mapboxIds: [47332003],
    geojson: fcGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Communication Studies and Journalism Building (CJ)',
    mapboxIds: [47332007],
    geojson: cjGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 08:00-18:00',
  },
  {
    id: 'Saint Ignatius of Loyola Church (SI)',
    mapboxIds: [47331997],
    geojson: siGeojson,
    addr: '4455 Rue West Broadway, Montréal, QC H4B 2A7',
    opening_hours: 'Mo-Fr 08:00-16:00; Sa 15:45-17:30; Su 10:00-11:00',
  },
  {
    id: 'Hingston Hall (HC)',
    mapboxIds: [604324438],
    geojson: hcGeojson,
    addr: 'Hingston Hall C, Montreal, Quebec H4B, Canada',
    opening_hours: '24/7',
  },
  {
    id: 'Hingston Hall (HB)',
    mapboxIds: [17887804],
    geojson: hbGeojson,
    addr: 'Hingston Hall B, Montreal, Quebec H4B, Canada',
    opening_hours: '24/7',
  },
  {
    id: 'Hingston Hall (HA)',
    mapboxIds: [17887803],
    geojson: haGeojson,
    addr: 'Hingston Hall A, Montreal, Quebec H4B, Canada',
    opening_hours: '24/7',
  },
  {
    id: 'Applied Science Hub (HU)',
    mapboxIds: [795012497],
    geojson: huGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Richard J. Renaud Science Complex (SP)',
    mapboxIds: [47331993],
    geojson: spGeojson,
    addr: '3475 Rue West Broadway, Montréal, QC H4B 2A7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Centre for Structural and Functional Genomics (GE)',
    mapboxIds: [545014554],
    geojson: geGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
];

const indoorMaps: IndoorMap[] = rawIndoorMaps.map((rawIndoorMap) => {
  const { bounds, levelsRange } = GeojsonService.extractLevelsRangeAndBounds(
    rawIndoorMap.geojson as IndoorMapGeoJSON
  );
  const indoorMap: IndoorMap = {
    id: rawIndoorMap.id,
    mapboxIds: rawIndoorMap.mapboxIds,
    bounds: bounds,
    geojson: rawIndoorMap.geojson as IndoorMapGeoJSON,
    levelsRange: levelsRange,
    addr: rawIndoorMap.addr,
    opening_hours: rawIndoorMap.opening_hours,
  };
  const rooms = getAllRooms(indoorMap.geojson);
  for (const room of rooms) {
    const roomRef = room.properties?.ref;
    if (typeof roomRef !== 'string') {
      continue;
    }
    LocalLocations.getInstance().add(roomRef, (name: string) => {
      return {
        name,
        coordinates: center(room).geometry.coordinates,
        data: {
          address: indoorMap.id,
          level: parseFloat(room.properties?.level),
          indoorMap: indoorMap,
          ref: roomRef,
          feature: room,
        },
      };
    });
  }
  return indoorMap;
});

const campusMapboxIds = indoorMaps.flatMap((indoorMap) => indoorMap.mapboxIds);

export { indoorMaps, campusMapboxIds };
