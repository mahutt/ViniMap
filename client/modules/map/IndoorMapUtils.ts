import type { BBox, Feature, LineString, Polygon } from 'geojson';
import {
  Coordinates,
  ExpressionSpecification,
  Level,
  Location,
  IndoorMap,
  LevelsRange,
} from './Types';
import GeojsonService from '@/services/GeojsonService';
import booleanPointInPolygon from '@turf/boolean-point-in-polygon';

export function overlap(bounds1: BBox, bounds2: BBox) {
  // If one rectangle is on left side of other
  if (bounds1[0] > bounds2[2] || bounds2[0] > bounds1[2]) {
    return false;
  }

  // If one rectangle is above other
  if (bounds1[3] < bounds2[1] || bounds2[3] < bounds1[1]) {
    return false;
  }

  return true;
}

export function filterWithLevel(
  initialFilter: ExpressionSpecification,
  level: Level,
  showFeaturesWithEmptyLevel: boolean = false
): ExpressionSpecification {
  return [
    'all',
    initialFilter,
    [
      'any',
      showFeaturesWithEmptyLevel ? ['!', ['has', 'level']] : false,
      [
        'all',
        ['has', 'level'],
        [
          'any',
          ['==', ['get', 'level'], level.toString()],
          [
            'all',
            ['!=', ['index-of', ';', ['get', 'level']], -1],
            [
              '>=',
              level,
              ['to-number', ['slice', ['get', 'level'], 0, ['index-of', ';', ['get', 'level']]]],
            ],
            [
              '<=',
              level,
              [
                'to-number',
                ['slice', ['get', 'level'], ['+', ['index-of', ';', ['get', 'level']], 1]],
              ],
            ],
          ],
        ],
      ],
    ],
  ];
}

export function bboxCenter(bbox: BBox): Coordinates {
  const [west, south, east, north] = bbox;
  return [(west + east) / 2, (south + north) / 2];
}

export function footwaysForLevel(indoorMap: IndoorMap, level: Level): Feature<LineString>[] {
  const footwayFeatures = indoorMap.geojson.features.filter(
    (feature) =>
      parseFloat(feature.properties?.level) === level &&
      feature.properties?.highway === 'footway' &&
      feature.geometry.type === 'LineString'
  );
  return footwayFeatures as Feature<LineString>[];
}

export function getIndoorFeatureFromCoordinates(
  indoorMap: IndoorMap,
  coordinates: Coordinates,
  level: Level
): Location | null {
  if (
    coordinates[0] < indoorMap.bounds[0] ||
    coordinates[0] > indoorMap.bounds[2] ||
    coordinates[1] < indoorMap.bounds[1] ||
    coordinates[1] > indoorMap.bounds[3] ||
    level < indoorMap.levelsRange.min ||
    level > indoorMap.levelsRange.max
  ) {
    return null;
  }

  for (let feature of indoorMap.geojson.features) {
    const featureLevel = GeojsonService.extractLevelFromFeature(feature);
    if (
      featureLevel === null ||
      (typeof featureLevel === 'number' && featureLevel !== level) ||
      (typeof featureLevel === 'object' &&
        (level < featureLevel.min || level > featureLevel.max)) ||
      feature.geometry.type !== 'Polygon' // Currently, we only intercept taps on polygon features
    ) {
      continue;
    }
    if (booleanPointInPolygon(coordinates, feature.geometry)) {
      const name =
        feature?.properties?.name ||
        getFallbackNameByAmenity(feature?.properties?.amenity) ||
        feature?.properties?.ref ||
        'Unknown room';
      return {
        coordinates,
        name,
        data: {
          address: indoorMap.id,
          isOpen: false,
          level: featureLevel,
          indoorMap: indoorMap,
          ref: feature?.properties?.ref,
          feature,
        },
      };
    }
  }

  return null;
}

const FALLBACK_NAME_BY_AMENITY = {
  toilets: 'Toilets',
  vending_machine: 'Vending machine',
  fast_food: 'Fast food',
  restaurant: 'Restaurant',
  cafe: 'Cafe',
  fountain: 'Fountain',
  eating_area: 'Eating area',
  information: 'Information',
};

const getFallbackNameByAmenity = (amenity: string | null) => {
  if (amenity && amenity in FALLBACK_NAME_BY_AMENITY) {
    return FALLBACK_NAME_BY_AMENITY[amenity as keyof typeof FALLBACK_NAME_BY_AMENITY];
  }
  return null;
};

// Given two features,
// returns the closest and smallest levels between the two features.
export const getStartEndLevels = (
  startFeature: Feature,
  endFeature: Feature
): [Level, Level] | null => {
  let startLevelOrRange = GeojsonService.extractLevelFromFeature(startFeature);
  let endLevelOrRange = GeojsonService.extractLevelFromFeature(endFeature);

  // If either the start or the end feature doesn't have a level,
  // we can't find a route between them.
  if (startLevelOrRange === null || endLevelOrRange === null) {
    return null;
  }

  let startLevel: Level;
  let endLevel: Level;

  // If the start feature spans a range of levels
  // and the end feature also spans a range of levels,
  // we provide a route between the two "closest" levels (minimize up/down stairs)
  if (typeof startLevelOrRange === 'object' && typeof endLevelOrRange === 'object') {
    [startLevel, endLevel] = getClosestLevels(startLevelOrRange, endLevelOrRange);
  }
  // If only the start feature spans a range of levels,
  // we provide a route from the start level closest to the end level
  else if (typeof startLevelOrRange === 'object') {
    endLevel = endLevelOrRange as Level;
    startLevel = getClosestLevel(endLevel, startLevelOrRange);
  }
  // If only the end feature spans a range of levels,
  // we provide a route to the end level closest to the start level
  else if (typeof endLevelOrRange === 'object') {
    startLevel = startLevelOrRange;
    endLevel = getClosestLevel(startLevel, endLevelOrRange);
  } else {
    startLevel = startLevelOrRange;
    endLevel = endLevelOrRange;
  }

  return [startLevel, endLevel];
};

// Given two levels ranges,
// returns one level from each range that are the closest to each other.
export const getClosestLevels = (
  levelsRange1: LevelsRange,
  levelsRange2: LevelsRange
): [Level, Level] => {
  let level1: Level;
  let level2: Level;

  if (levelsRange1.min <= levelsRange2.max && levelsRange1.max >= levelsRange2.min) {
    const smallestCommonLevel = Math.max(levelsRange1.min, levelsRange2.min);
    level1 = smallestCommonLevel;
    level2 = smallestCommonLevel;
  } else {
    level1 = levelsRange1.min > levelsRange2.max ? levelsRange1.min : levelsRange1.max;
    level2 = levelsRange2.max < level1 ? levelsRange2.max : levelsRange2.min;
  }
  return [level1, level2];
};

// Given a level and a levels range,
// returns the closest level from the range to the given level.
export const getClosestLevel = (level: Level, levelsRange: LevelsRange): Level => {
  if (level < levelsRange.min) {
    return levelsRange.min;
  } else if (level > levelsRange.max) {
    return levelsRange.max;
  } else {
    return level;
  }
};

// Returns an array of features that connect two levels of an indoor map.
// Currently, only one connection is returned. We will eventually return multiple connections
// to support more complex indoor maps.
export const getConnectionsBetween = (
  startLevel: Level,
  endLevel: Level,
  indoorMap: IndoorMap,
  mode: string
): Feature<Polygon>[] => {
  const temp = startLevel;
  startLevel = endLevel;
  endLevel = temp;

  //mode = 'handicap';
  console.log(mode);

  const getPathForDisabled = mode === 'handicap' ? true : false;
  let possibleConnections = indoorMap.geojson.features.filter(
    (feature) =>
      (feature.properties?.highway === 'elevator' && getPathForDisabled) ||
      ((feature.properties?.stairs === 'yes' || feature.properties?.conveying === 'yes') &&
        !getPathForDisabled)
  );

  const usableConnections = possibleConnections.filter((feature) => {
    if (feature.geometry.type !== 'Polygon') {
      return false;
    }

    const level = GeojsonService.extractLevelFromFeature(feature);
    if (level !== null && typeof level === 'object') {
      if (startLevel < endLevel) {
        return level.min <= startLevel && level.max >= endLevel;
      } else {
        return level.min <= endLevel && level.max >= startLevel;
      }
    }

    return false;
  });

  if (usableConnections.length == 0) {
    return getDisjointConnections(startLevel, endLevel, possibleConnections as Feature<Polygon>[]);
  }

  return usableConnections.slice(0, 1) as Feature<Polygon>[];
};

export const getDisjointConnections = (
  startLevel: Level,
  endLevel: Level,
  possibleConnections: Feature<Polygon>[]
): Feature<Polygon>[] => {
  const usableConnections = [];
  const first_connection = possibleConnections.filter((feature) => {
    if (feature.geometry.type !== 'Polygon') {
      return false;
    }
    const level = GeojsonService.extractLevelFromFeature(feature);
    if (level !== null && typeof level === 'object') {
      return startLevel <= level.max && startLevel >= level.min;
    }
  });
  //Didn't find any possible connections containing the startLevel
  if (first_connection.length == 0) {
    return [];
  }

  //First Usable Connection. We loop until we've found enough usable connections to complete the route
  let current_connection = first_connection[0];
  usableConnections.push(current_connection);

  //Loop to get all pieces. Similar to linked list
  while (true) {
    const current_level = GeojsonService.extractLevelFromFeature(current_connection);
    let current_max: Level;
    if (current_level !== null && typeof current_level === 'object') {
      current_max = current_level.max;
    }

    const next_conection = possibleConnections.filter((feature) => {
      if (feature.geometry.type !== 'Polygon') {
        return false;
      }
      const next_connection_level = GeojsonService.extractLevelFromFeature(feature);
      if (next_connection_level !== null && typeof next_connection_level === 'object') {
        return next_connection_level.min == current_max;
      }
    });

    if (next_conection.length == 0) {
      return [];
    }

    const next_levels = GeojsonService.extractLevelFromFeature(next_conection[0]);
    let next_levels_max = -100;
    if (next_levels !== null && typeof next_levels === 'object') {
      next_levels_max = next_levels.max;
    }

    usableConnections.push(next_conection[0]);
    if (next_levels_max >= endLevel) {
      break;
    } else {
      current_connection = next_conection[0];
      continue;
    }
  }
  // console.log('AFTER LOOP');
  // usableConnections.forEach((F) => console.log(JSON.stringify(F)));
  return usableConnections.reverse() as Feature<Polygon>[];
};
