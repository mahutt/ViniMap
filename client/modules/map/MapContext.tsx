import React, { createContext, useContext, useRef, useState, useMemo } from 'react';
import Mapbox from '@rnmapbox/maps';
import { getRoute } from './MapService';
import { Coordinates } from './Types';

export interface Location {
  name: string | null;
  coordinates: Coordinates;
  data?: any;
}

export enum MapState {
  Idle,
  RoutePlanning,
  Information,
  SelectingStartLocation,
  SelectingEndLocation,
}

const DEFAULT_COORDINATES: Coordinates = [-73.5789, 45.4973];

type MapContextType = {
  mapRef: React.RefObject<Mapbox.MapView>;
  cameraRef: React.RefObject<Mapbox.Camera>;
  centerCoordinate: [number, number];
  zoomLevel: number;

  pitchLevel: number;
  setPitchLevel: (pitchLevel: number) => void;

  state: MapState;
  startLocation: Location | null;
  endLocation: Location | null;
  routeCoordinates: Coordinates[];
  setCenterCoordinate: (centerCoordinate: [number, number]) => void;
  setZoomLevel: (zoomLevel: number) => void;
  setState: (state: MapState) => void;
  setStartLocation: (startLocation: Location | null) => void;
  setEndLocation: (endLocation: Location | null) => void;
  flyTo: (coords: [number, number], zoomLevel?: number) => void;
  loadRouteFromCoordinates: (
    startCoordinates: Coordinates,
    endCoordinates: Coordinates,
    mode?: string
  ) => Promise<void>;
};

const MapContext = createContext<MapContextType | undefined>(undefined);

export const MapProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const mapRef = useRef<Mapbox.MapView | null>(null);
  const cameraRef = useRef<Mapbox.Camera | null>(null);
  const [centerCoordinate, setCenterCoordinate] = useState<[number, number]>(DEFAULT_COORDINATES);
  const [zoomLevel, setZoomLevel] = useState(15);
  const [pitchLevel, setPitchLevel] = useState(0);
  const [state, setState] = useState<MapState>(MapState.Idle);

  const setStateWithLogging = (newState: MapState) => {
    switch (newState) {
      case MapState.SelectingStartLocation:
        setStartLocation(null);
        break;
      case MapState.RoutePlanning:
        break;
    }

    setState(newState);
  };

  const [startLocation, setStartLocation] = useState<Location | null>(null);
  const [endLocation, setEndLocation] = useState<Location | null>(null);
  const [routeCoordinates, setRouteCoordinates] = useState<Coordinates[]>([]);

  const flyTo = useMemo(
    () => (newCenterCoordinate: [number, number], newZoomLevel?: number) => {
      if (cameraRef.current) {
        cameraRef.current.setCamera({
          centerCoordinate: newCenterCoordinate,
          zoomLevel: newZoomLevel ?? zoomLevel,
          animationDuration: 2000,
          pitch: pitchLevel,
        });
        setCenterCoordinate(newCenterCoordinate);
        if (newZoomLevel) setZoomLevel(newZoomLevel);
      }
    },
    [zoomLevel, pitchLevel]
  );

  const loadRouteFromCoordinates = async (
    startCoordinates: Coordinates,
    endCoordinates: Coordinates,
    mode = 'walking'
  ): Promise<void> => {
    if (!startCoordinates || !endCoordinates) {
      return;
    }
    return getRoute(startCoordinates, endCoordinates, mode)
      .then((data) => {
        if (data?.coordinates) {
          if (data.coordinates.length > 0) {
            flyTo(data.coordinates[0], zoomLevel);
          }
          setRouteCoordinates(data.coordinates);
        }
      })
      .catch((error) => {
        console.error('Error setting route:', error);
      });
  };

  const value = useMemo(
    () => ({
      mapRef,
      cameraRef,
      centerCoordinate,
      zoomLevel,
      pitchLevel,
      setPitchLevel,
      state,
      startLocation,
      endLocation,
      routeCoordinates,
      setCenterCoordinate,
      setZoomLevel,
      setState: setStateWithLogging,
      setStartLocation,
      setEndLocation,
      flyTo,
      loadRouteFromCoordinates,
    }),
    [
      centerCoordinate,
      zoomLevel,
      state,
      startLocation,
      endLocation,
      routeCoordinates,
      flyTo,
      pitchLevel,
    ]
  );

  return <MapContext.Provider value={value}>{children}</MapContext.Provider>;
};

export const useMap = () => {
  const context = useContext(MapContext);
  if (context === undefined) {
    throw new Error('useMap must be used within a MapProvider');
  }
  return context;
};
export { Coordinates };
