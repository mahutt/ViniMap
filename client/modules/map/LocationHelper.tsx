import { Location } from './MapContext';
import CoordinateService from '@/Services/CoordinateService';

export const getCurrentLocationAsStart = async (setLocation: (location: Location) => void) => {
  try {
    console.log('getCurrentLocationAsStart: Attempting to get coordinates');
    const tempCoordinates = await CoordinateService.getCurrentCoordinates();

    console.log('Coordinates retrieved:', tempCoordinates);

    if (tempCoordinates) {
      console.log('Setting location with coordinates:', tempCoordinates);
      setLocation({
        name: 'Current location',
        coordinates: tempCoordinates,
      });
      console.log('Location set successfully');
    } else {
      console.log('No coordinates available');
    }
  } catch (error) {
    console.error('Detailed error in getCurrentLocationAsStart:', error);
  }
};
