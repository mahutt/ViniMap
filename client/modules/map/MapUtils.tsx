import { Coordinates } from './MapContext';

/**
 * Calculates the Euclidean distance between two points.
 *
 * @param start - The starting coordinates as a tuple [x, y].
 * @param end - The ending coordinates as a tuple [x, y].
 * @returns The Euclidean distance between the start and end coordinates.
 */
const calculateEuclideanDistance = (start: Coordinates, end: Coordinates): number => {
  return Math.sqrt(Math.pow(start[0] - end[0], 2) + Math.pow(start[1] - end[1], 2));
};

export { calculateEuclideanDistance };
