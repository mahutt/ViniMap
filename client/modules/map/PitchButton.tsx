import { StyleSheet,Button,View, TouchableOpacity,Text} from 'react-native';
import Mapbox from '@rnmapbox/maps';
import React from 'react';
import { useMap } from './MapContext';
import MaterialCommunityIcons from '@expo/vector-icons/MaterialCommunityIcons';


export default function PitchButton() {
  const {  pitchLevel,setPitchLevel ,setZoomLevel,zoomLevel,cameraRef,centerCoordinate, mapRef} = useMap();

  const pressHandler =async  () =>{
    const currentZoom = await mapRef.current?.getZoom() as number
    setZoomLevel(currentZoom )
    const newPitch =(pitchLevel == 75)?0:75
    setPitchLevel( newPitch) 
   
  
  }
  return (
    <View style={styles.pitchButtonContainer}>
        <TouchableOpacity onPress={ pressHandler }><MaterialCommunityIcons name="angle-obtuse" size={35} color="white"  /></TouchableOpacity>
    
    </View>
  );
}

const styles = StyleSheet.create({
  pitchButtonContainer: {
   position:'absolute',
   zIndex:100,
   right:'4%',
   top:'92%',
   backgroundColor:'#912338',
   borderRadius:50,
   height:50,
   width:50,

  justifyContent:'center',
  alignItems:'center',
  opacity:0.89,
  

  
    
  },
  
 
  
});
