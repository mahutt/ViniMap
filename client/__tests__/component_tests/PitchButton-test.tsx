import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { TouchableOpacity } from 'react-native';
import PitchButton from '@/modules/map/PitchButton';
import { useMap } from '@/modules/map/MapContext';

// Mock the MapContext hook
jest.mock('@/modules/map/MapContext', () => ({
  useMap: jest.fn(),
}));

// Mock the MaterialCommunityIcons component
jest.mock('@expo/vector-icons/MaterialCommunityIcons', () => 'MockedMaterialCommunityIcons');

describe('PitchButton', () => {
  const mockSetPitchLevel = jest.fn();
  const mockSetZoomLevel = jest.fn();
  const mockSetCenterCoordinate = jest.fn();
  const mockGetZoom = jest.fn().mockResolvedValue(10);
  const mockGetCenter = jest.fn().mockResolvedValue([45.5, -73.6]);
  const mockMapRef = {
    current: {
      getZoom: mockGetZoom,
      getCenter: mockGetCenter,
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();

    // Default mock implementation
    (useMap as jest.Mock).mockReturnValue({
      pitchLevel: 0,
      setPitchLevel: mockSetPitchLevel,
      setZoomLevel: mockSetZoomLevel,
      setCenterCoordinate: mockSetCenterCoordinate,
      mapRef: mockMapRef,
    });
  });

  it('renders correctly', () => {
    const { toJSON } = render(<PitchButton />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('toggles pitch from 0 to 75 when pressed', async () => {
    (useMap as jest.Mock).mockReturnValue({
      pitchLevel: 0,
      setPitchLevel: mockSetPitchLevel,
      setZoomLevel: mockSetZoomLevel,
      setCenterCoordinate: mockSetCenterCoordinate,
      mapRef: mockMapRef,
    });

    const { UNSAFE_getAllByType } = render(<PitchButton />);
    const touchableOpacity = UNSAFE_getAllByType(TouchableOpacity)[0];

    fireEvent.press(touchableOpacity);

    await waitFor(() => {
      expect(mockGetCenter).toHaveBeenCalled();
      expect(mockGetZoom).toHaveBeenCalled();
      expect(mockSetCenterCoordinate).toHaveBeenCalledWith([45.5, -73.6]);
      expect(mockSetZoomLevel).toHaveBeenCalledWith(10);
      expect(mockSetPitchLevel).toHaveBeenCalledWith(75);
    });
  });

  it('toggles pitch from 75 to 0 when pressed', async () => {
    (useMap as jest.Mock).mockReturnValue({
      pitchLevel: 75,
      setPitchLevel: mockSetPitchLevel,
      setZoomLevel: mockSetZoomLevel,
      setCenterCoordinate: mockSetCenterCoordinate,
      mapRef: mockMapRef,
    });

    const { UNSAFE_getAllByType } = render(<PitchButton />);
    const touchableOpacity = UNSAFE_getAllByType(TouchableOpacity)[0];

    fireEvent.press(touchableOpacity);

    await waitFor(() => {
      expect(mockSetPitchLevel).toHaveBeenCalledWith(0);
    });
  });

  it('handles null mapRef gracefully', async () => {
    (useMap as jest.Mock).mockReturnValue({
      pitchLevel: 0,
      setPitchLevel: mockSetPitchLevel,
      setZoomLevel: mockSetZoomLevel,
      setCenterCoordinate: mockSetCenterCoordinate,
      mapRef: { current: null },
    });

    const { UNSAFE_getAllByType } = render(<PitchButton />);
    const touchableOpacity = UNSAFE_getAllByType(TouchableOpacity)[0];

    fireEvent.press(touchableOpacity);

    await waitFor(() => {
      expect(mockSetPitchLevel).toHaveBeenCalledWith(75);
    });
  });
});
