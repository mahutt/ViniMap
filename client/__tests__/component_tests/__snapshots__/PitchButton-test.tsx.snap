// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PitchButton renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#912338",
      "borderRadius": 50,
      "height": 50,
      "justifyContent": "center",
      "opacity": 0.89,
      "position": "absolute",
      "right": 36,
      "top": 130,
      "width": 50,
      "zIndex": 2,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <MockedMaterialCommunityIcons
      color="white"
      name="angle-obtuse"
      size={35}
    />
  </View>
</View>
`;
